packages/commands/src/components/command-menu.js

```diff
@@ -0,0 +1,198 @@
+/**
+ * External dependencies
+ */
+import { Command } from 'cmdk';
+
+/**
+ * WordPress dependencies
+ */
+import { useSelect } from '@wordpress/data';
+import { useState, useEffect, useRef, useCallback } from '@wordpress/element';
+import { __ } from '@wordpress/i18n';
+import { Modal, TextHighlight } from '@wordpress/components';
+
+/**
+ * Internal dependencies
+ */
+import { store as commandsStore } from '../store';
+
+function CommandMenuLoader( { name, search, hook, setLoader, close } ) {
+	const { isLoading, commands = [] } = hook( { search } ) ?? {};
+	useEffect( () => {
+		setLoader( name, isLoading );
+	}, [ setLoader, name, isLoading ] );
+
+	return (
+		<>
+			<Command.List>
+				{ isLoading && (
+					<Command.Loading>{ __( 'Searching…' ) }</Command.Loading>
+				) }
+
+				{ commands.map( ( command ) => (
+					<Command.Item
+						key={ command.name }
+						value={ command.label ?? command.name }
+						onSelect={ () => command.callback( { close } ) }
+					>
+						<span className="commands-command-menu__item">
+							<TextHighlight
+								text={ command.label }
+								highlight={ search }
+							/>
+						</span>
+					</Command.Item>
+				) ) }
+			</Command.List>
+		</>
+	);
+}
+
+export function CommandMenuLoaderWrapper( { hook, search, setLoader, close } ) {
+	// loader is actually a custom hook
+	// so to avoid breaking the rules of hooks
+	// the CommandsPerPage component need to be
+	// remounted on each loader change
+	// We use the key state to make sure we do that properly.
```

Comment:
A few notes:

1. `loader` is mentioned in the comment but is not a binding. I assume it means `hook`. For consistency, probably change everything to `loader` + `setLoader`.
2. `CommandsPerPage` is mentioned in the comment but doesn't match anything. Is it `CommandMenuLoader`?
3. This feels a bit like a code smell. Is there a better approach out there that doesn't rely on user-defined hooks? Maybe not.

Reply:
> CommandsPerPage is mentioned in the comment but doesn't match anything. Is it CommandMenuLoader?

yes, that has been removed.

Reply:
> This feels a bit like a code smell. Is there a better approach out there that doesn't rely on user-defined hooks? Maybe not.

Agreed it's not perfect, we already have the same in "useAutocomplete". But to be honest I can't think of anything better.

I'd like for the command loader to be dynamic (be able to use "search" to filter results). One option would be to make it an async function but then you lose a lot of things (not able to use useSelect...).

Reply:
I'm conflicted about this too, but I don't have an alternative besides `async` functions.